Fix: Unknown nested elements in XML packages elements (like Host etc.) will cause erasure of settings

Delwing opened the pull request:
<!-- Keep the title short & concise so anyone non-technical can understand it,
     the title appears in PTB changelogs -->
#### Brief overview of PR changes/additions

Having unknown elements in XML, might cause clearing of already properly set elements. This shouldn't happen, as we're flattening Host package, but still good to fix that issue.
Example:

```xml
<Host autoClearCommandLineAfterSend="no" HighlightHistory="yes" printCommand="yes" USE_IRE_DRIVER_BUGFIX="no" mUSE_FORCE_LF_AFTER_PROMPT="no" mUSE_UNIX_EOL="no" mNoAntiAlias="no" mEchoLuaErrors="yes" runAllKeyMatches="no" AmbigousWidthGlyphsToBeWide="auto" mRawStreamDump="yes" mIsLoggingTimestamps="no" ...>
			<name>Test</name>
			<mInstalledPackages />
			...
			<unknowns>
				<unknown>1</unknown>
				<unknown>2</unknown>
			</unknowns>
			<stopwatches />
		</Host>
```
initially will set all values correctly in Host object, second run will happen, like those values are never set, because it will treat `unknowns` node just like it was `Host` node.

#### Motivation for adding to Mudlet
Fixing

#### Other info (issues closed, discussion etc)

Same applies to other elements, like Triggers, Aliases etc.

#### Release post highlight
<!--
Use this space if you wish to write a short statement or example for inclusion
in the release post for the next release. 
-->

Delwing committed:
Fix: Clearing of previously set elements on nested unknown elements

vadi2 commented:
@wiploo @mpconley could you test using this version instead of PTB, and then switching back to release?

wiploo commented:
Unfortuntelly I have the same problem with this PR. I lost all profile settings.

SlySven reviewed (commented):
I think changing all the things to the `readUnknownXxxx()` methods just causes a further warning about unknown elements (and does not parse them) whereas the `readXxxx()` goes on to read (recursively!) the next valid element. As such this would cause more data loss...

Delwing commented:
@SlySven unless I misunderstood something here, I'd say this is desired...

While `/MudletPackage/HostPackage/Host/wrapAt` is correct place to check then some accidental `/MudletPackage/HostPackage/Host/DummyWindow/wrapAt`  because we don't know `DummyWindow` node, we should not care what's inside, it might contain values that will override our correct values. Not very probable. that they will contain anything valid, but still.

Delwing commented:
> Unfortuntelly I have the same problem with this PR. I lost all profile settings.

Just to be clear. It can fix things between PTB AND release branch 4.15, not the current version 4.14.1. Did you switch between those two? Switching to 4.14.1 will still cause erasure.

vadi2 commented:
> Switching to 4.14.1 will still cause erasure.

We can't have that, not in Mudlet anyway - losing people's stuff isn't OK

Delwing commented:
> > Switching to 4.14.1 will still cause erasure.
> 
> We can't have that, not in Mudlet anyway - losing people's stuff isn't OK

Once again, I can't fix previous versions.

SlySven commented:
> While /MudletPackage/HostPackage/Host/wrapAt is correct place to check then some accidental /MudletPackage/HostPackage/Host/DummyWindow/wrapAt because we don't know DummyWindow node, we should not care what's inside, it might contain values that will override our correct values. Not very probable. that they will contain anything valid, but still.

Ah, yeah, I get you, in that case there is a **[`(void) QXmlStreamReader::skipCurrentElement()`](https://doc.qt.io/qt-5/qxmlstreamreader.html#skipCurrentElement)** method that should be called when the parser is on the ***startElement*** of the unknown item (I believe it is still possible to read the name and *attributes* of that item at that point) so as to skip the children and move the parser to the matching ***endElement*** ready to go around that `while {...}` loop again and then return...

Delwing commented:
Sure, might be wise to just skip it entirely and not produce further unknown lines.

Delwing committed:
Fix: ignore unknown elements children

Delwing committed:
Fix: correct warning messages

SlySven commented on src/XMLimport.cpp:
```diff
@@ -611,144 +611,112 @@ void XMLimport::readHelpPackage()
 {
     while (!atEnd()) {
-        readNext();
+        skipCurrentElement();
```
I think this has to be *after* the debug message - as it advances the parser off of the *current* element which you are skipping and leaves it with the parent item being the current one so the details printed out would not be what you think it is...

TBH I have had fun with the XML reader in the past - and the best way to work out what is happening is to single step through the loop in the debugger with watches on `name()`, `attributes()`, `text()` and `tokenType()` whilst also having the XML file open in a suitable editor/viewer...

Delwing replied:
I've checked. This is correct.
```xml
<unknown>
  <inner></inner>
</unknown>
```

In this case we're outputting `unknow`, not `inner` and that's desired.

vadi2 replied:
@SlySven could you confirm?

vadi2 commented:
It doesn't look like we got around to testing this - I don't want to merge this last-minute into the update, let's cover it in the next update.

SlySven committed:
Merge branch 'development' into fix-unknown-element-erasure

SlySven committed:
Merge branch 'development' into fix-unknown-element-erasure

ghost commented:
<details>
<summary><b>👇 <a href="https://app.codesee.io/r/reviews?pr=5895&src=https%3A%2F%2Fgithub.com%2FMudlet%2FMudlet">Click on the image</a> for a new way to code review</b>
</summary>

  - Make big changes easier — review code in small groups of related files
  - Know where to start — see the whole change at a glance
  - Take a code tour — explore the change with an interactive tour
  - Make comments and review — all fully sync’ed with github

    [Try it now!](https://app.codesee.io/r/reviews?pr=5895&src=https%3A%2F%2Fgithub.com%2FMudlet%2FMudlet)
</details>

[![Review these changes using an interactive CodeSee Map](https://s3.us-east-2.amazonaws.com/maps.codesee.io/images/github/Mudlet/Mudlet/5895/7030e212/90fc6cc18d1e19412c711dd2fe3fd56945de59f9.svg)](https://app.codesee.io/r/reviews?pr=5895&src=https%3A%2F%2Fgithub.com%2FMudlet%2FMudlet)

#### Legend
![CodeSee Map Legend](https://app.codesee.io/Legend_v2.png)

SlySven committed:
Merge branch 'development' into fix-unknown-element-erasure

SlySven committed:
Merge branch 'development' into fix-unknown-element-erasure

SlySven commented:
I have tested the current code and when I fed it a doctored game save file with the following unexpected ("badElement") data inserted **under** the `<host>` element (each line of text ends with a line-feed and then multiple tabs {3 or 4} indent the text):
```
			<mInstalledPackages>
				<string>deleteOldProfiles</string>
				<string>echo</string>
				<string>run-lua-code-v5</string>
				<string>eventTest_labelDeletion</string>
				<string>NonLatin1Tests</string>
			</mInstalledPackages>
			<badElement type="Not good" ignorable="Oh yeah!">
				<string>this is not supposed to be read</string>
				<string>nor this string</string>
				<integer>42</integer>
			</badElement>
			<url>34.217.254.118</url>
			<serverPackageName>nothing</serverPackageName>
			<serverPackageVersion>-1</serverPackageVersion>
```
this provoked the following command line debug output from Mudlet:
```
[LOADING PROFILE]: "/home/stephen/.config/mudlet/profiles/WoTMUD_Jomin/current/bad_elements.xml"
XMLimport::readUnknownHostElement() ERROR: UNKNOWN Host Package Element, name: "" and content: "\n\t\t\t\t"
XMLimport::readUnknownHostElement() ERROR: UNKNOWN Host Package Element, name: "string" and content: ""
XMLimport::readUnknownHostElement() ERROR: UNKNOWN Host Package Element, name: "" and content: "\n\t\t\t\t"
XMLimport::readUnknownHostElement() ERROR: UNKNOWN Host Package Element, name: "string" and content: ""
XMLimport::readUnknownHostElement() ERROR: UNKNOWN Host Package Element, name: "" and content: "\n\t\t\t\t"
XMLimport::readUnknownHostElement() ERROR: UNKNOWN Host Package Element, name: "integer" and content: ""
XMLimport::readUnknownHostElement() ERROR: UNKNOWN Host Package Element, name: "" and content: "\n\t\t\t"
XMLimport::readUnknownHostElement() ERROR: UNKNOWN Host Package Element, name: "badElement" and content: ""
```

So I don't think https://github.com/Mudlet/Mudlet/pull/5895#discussion_r810469114 is exactly correct...

SlySven committed:
Fix: unknown element erasure

This should give more useful feed-back when unexpected elements are
encountered in game save files and Mudlet item imports from a file or
clipboard. It refactors multiple near-copies of some reporting code to a
common method and extends the functionality to also cover Variables in such
files (which can be set to be saved from the "Variables" view in the
Editor). It also reports the content of the unexpected element better as
it does NOT get caught up with reporting the white-space characters that we
leave between elements in our XML files rather than the meaningful, but
unexpected content.

Rename to better reflect the name of the XML element they parse (though
both "Item" and "ItemGroup" are both parsed by the same method "readItem"):
* `(void) XMLimport::readHostPackage(Host*)`
                                           ==> `XMLimport::readHost(Host*)`
* `(int) XMLimport::readTriggerGroup(TTrigger*)`
                                    ==> `XMLimport::readTrigger(TTrigger*)`
* `(int) XMLimport::readTimerGroup(TTimer*)`
                                        ==> `XMLimport::readTimer(TTimer*)`
* `(int) XMLimport::readAliasGroup(TAlias*)`
                                        ==> `XMLimport::readAlias(TAlias*)`
* `(int) XMLimport::readActionGroup(TAction*)`
                                      ==> `XMLimport::readAction(TAction*)`
* `(int) XMLimport::readScriptGroup(TScript*)`
                                      ==> `XMLimport::readScript(TScript*)`
* `(int) XMLimport::readKeyGroup(TKey*)`    ==> `XMLimport::readKey(TKey*)`

* `(void) XMLimport::readVariableGroup(TVar*)`
                                       ==> `XMLimport::readVariable(TVar*)`

Delete, and replace with calls to unifying:
 `(void) XMLimport::readUnknownElement(const QString&)`:

* `(void) XMLimport::readUnknownPackage()`
* `(void) XMLimport::readUnknownHostElement()`
* `(void) XMLimport::readUnknownTriggerElement()`
* `(void) XMLimport::readUnknownTimerElement()`
* `(void) XMLimport::readUnknownAliasElement()`
* `(void) XMLimport::readUnknownActionElement()`
* `(void) XMLimport::readUnknownScriptElement()`
* `(void) XMLimport::readUnknownKeyElement()`

Signed-off-by: Stephen Lyons <slysven@virginmedia.com>

SlySven commented:
:warning: Please do not "update branch" until a decision on https://github.com/Delwing/Mudlet/pull/83 is taken... :crossed_fingers:

Delwing committed:
Merge pull request #83 from SlySven/Fix_unknownElementErasure

Fix: unknown element erasure

Delwing commented:
@SlySven merged that one of yours into that one - one Vadi's remark stayed unresolved, but since this is trivial thing to change if needed, I decided not to wait - fixing issue is most crucial here.

vadi2 commented:
@SlySven mind having another look at this?

SlySven commented:
Will do, as soon as I can (but that might be a day or two)...

SlySven committed:
Merge branch 'development' into Fix_unknownElementErasure_2

SlySven committed:
Revise: change format of debug message when reporting a file line & column

This was suggested by @vadi2 during peer review!

Signed-off-by: Stephen Lyons <slysven@virginmedia.com>

SlySven commented:
> @SlySven mind having another look at this?

Done as: https://github.com/Delwing/Mudlet/pull/87 it **needs merging into Delwing's PR by him**. Note that only 
https://github.com/Delwing/Mudlet/pull/87/commits/591b0e8d394a4f8cc9115d3d78d0d6f407d87e64 is the actual change - all the preceding ones are the result of merging in the changes to the `development` branch - and will "evaporate" (:crossed_fingers:) when this PR eventually (very quick one hopes) gets into the `development` branch...!

Delwing committed:
Merge pull request #87 from SlySven/Fix_unknownElementErasure_2

Fix: unknown element erasure 2

SlySven commented:
Okay, so @vadi2 has merged in something else into `development` whist I was putting that second PR together so it has slipped a little bit behind - hence the "This branch is out-of-date..." is still there. :roll_eyes: but it should be good to go now. :pleading_face:
